common/achievements.cpp

```diff
@@ -0,0 +1,178 @@
+ *
+ */
+
+
+#include "common/achievements.h"
+#include "common/config-manager.h"
+#include "common/debug.h"
+
+namespace Common {
+
+DECLARE_SINGLETON(AchievementsManager);
+
+
+AchievementsManager::AchievementsManager() {
+	unsetActiveDomain();
+}
+
+
+AchievementsManager::~AchievementsManager() {
+}
+
+bool AchievementsManager::setActiveDomain(AchievementsPlatform platform, const String &appId) {
+	String suffix = platform == STEAM_ACHIEVEMENTS ? "steam-" + appId : appId;
+	String achDomainId = "achievements-" + suffix;
+	String statDomainId = "statistics-" + suffix;
+
+	if (_achDomainId == achDomainId && _statDomainId == statDomainId) {
+		return true;
+	}
+
+	if (isReady()) {
+		unsetActiveDomain();
+	}
+
+	ConfMan.addMiscDomain(achDomainId);
+	ConfMan.addMiscDomain(statDomainId);
+
+	_achDomainId = achDomainId;
+	_statDomainId = statDomainId;
+	return true;
+}
+
+
+bool AchievementsManager::unsetActiveDomain() {
+	if (!isReady()) {
+		return true;
+	}
+
+	ConfMan.flushToDisk();
+
+	_achDomainId = "";
+	_statDomainId = "";
+	return true;
+}
+
+
+bool AchievementsManager::setAchievement(const String &id) {
+	if (!isReady()) {
+		return false;
+	}
+	if (isAchieved(id)) {
+		return true;
+	}
+
+	//TODO: insert a GUI notification here
+	debug("Achievement unlocked: %s", id.c_str());
+
+	ConfMan.setBool(id, true, _achDomainId);
+	ConfMan.flushToDisk();
+	return true;
+}
+
+
+bool AchievementsManager::isAchieved(const String &id) {
+	if (!isReady()) {
+		return false;
+	}
+
+	return ConfMan.hasKey(id, _achDomainId) && ConfMan.getBool(id, _achDomainId);
+}
+
+
+bool AchievementsManager::clearAchievement(const String &id) {
+	if (!isReady()) {
+		return false;
+	}
+
+	ConfMan.removeKey(id, _achDomainId);
+	ConfMan.flushToDisk();
+	return true;
+}
+
+
+bool AchievementsManager::setStatFloat(const String &id, float value) {
+	if (!isReady()) {
+		return false;
+	}
+
+	char tmp[32];
+	sprintf(tmp, "%8.8f", value);
+	ConfMan.set(id, tmp, _statDomainId);
+	ConfMan.flushToDisk();
+	return 0;
+}
+
+
+float AchievementsManager::getStatFloat(const String &id) {
+	if (!isReady()) {
+		return 0.0;
+	}
+
+	String tmp = ConfMan.get(id, _statDomainId);
+	return atof(tmp.c_str());
+}
+
+
+bool AchievementsManager::setStatInt(String const &id, int value) {
+	if (!isReady()) {
+		return false;
+	}
+
+	ConfMan.setInt(id, value, _statDomainId);
+	ConfMan.flushToDisk();
+	return 0;
+}
+
+
+int AchievementsManager::getStatInt(String const &id) {
+	if (!isReady()) {
+		return 0;
+	}
+
+	return ConfMan.getInt(id, _statDomainId);
+}
+
+
+bool AchievementsManager::resetAllAchievements() {
+	if (!isReady()) {
+		return false;
+	}
+
+	ConfMan.removeMiscDomain(_achDomainId);
+	ConfMan.flushToDisk();
+	return 0;
+}
+
+
+bool AchievementsManager::resetAllStats() {
+	if (!isReady()) {
+		return false;
+	}
+
+	ConfMan.removeMiscDomain(_statDomainId);
+	ConfMan.flushToDisk();
+	return 0;
+}
+
+
+} // End of namespace Common 
```

Comment:
No end of line

Reply:
Fixed at 67f82e741ca0540a747d0c7c8762bedecc9bcd88